### PR TITLE
Add Access control tests in lacking action classes #7431

### LINKED ISSUE #7431: Add Access control tests in lacking action classes
As discussed in #7332

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #7431

### CODE REVIEW

**Comment by sujeet14108** [comment]
Not implemented for all the classes yet but can anyone review till now so that i get to know that i am moving in right direction

**Comment by wkurniawan07** [comment]
I'm not sure what kind of assistance you need, but what you've done is adding missing access control tests which indeed is what the issue is about.

**Comment by sujeet14108** [comment]
@wkurniawan07 Can you help me with `AdminEmailComposePageActionTest` accesscontrol

**Comment by sujeet14108** [comment]
There are similar problem in certain other classes where different dataBundle is used other than `typicaldataBundle`

**Comment by whipermr5** [comment]
@sujeet14108 Maybe you need to load both databundles through the use of `super`, like in #7458?

**Comment by whipermr5** [comment]
@sujeet14108 Is this ready for review?

**Comment by sujeet14108** [comment]
Facing problem in `InstructorEditInstructorFeedbackPageActionTest` and `InstructorFeedbackEditCopyActionTest` would try to do it by today

**Comment by sujeet14108** [comment]
@whipermr5 can you help me with failing test.

**Comment by whipermr5** [comment]
@sujeet14108 Are you able to reproduce the error locally?

**Comment by sujeet14108** [comment]
@whipermr5 Yes error was in `InstructorFeedbackEditCopyActionTest` 

     java.lang.AssertionError: expected:<You are not registered in the course > but was:<>
	     at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:101)
	     at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:108)
	     at teammates.test.driver.AssertHelper.assertContains(AssertHelper.java:63)
	     at teammates.test.cases.action.BaseActionTest.verifyCannotAccess(BaseActionTest.java:662)
	     at teammates.test.cases.action.BaseActionTest.verifyUnaccessibleForStudents(BaseActionTest.java:576)`

**Comment by whipermr5** [comment]
@sujeet14108 Try to trace the problem and understand what's causing it.

**Comment by sujeet14108** [comment]
@whipermr5  Ready for review :)

File: src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java [hunk 1]
```diff
@@ -139,6 +140,21 @@ protected InstructorEditInstructorFeedbackPageAction getAction(String... params)
     @Override
     @Test
     protected void testAccessControl() throws Exception {
-        //TODO: implement this
+        dataBundle = getTypicalDataBundle();
+        removeAndRestoreDataBundle(dataBundle);
```
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java]: I have tried earlier by this way but it was was not working

Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java]: @sujeet14108 Do investigate and understand why it did not work, or update here if you already know :)

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java]: @whipermr5  In 

     @Override
    protected void prepareTestData() {
        super.prepareTestData();
        dataBundle = loadDataBundle("/InstructorEditStudentFeedbackPageTest.json");
        removeAndRestoreDataBundle(dataBundle);
    }

By `super.prepareTestData()` we can access the `TypicalDataBundle` by `super.databundle` but  in `testAccessControl`  we are calling the other base class methods like `verifyUnaccessibleWithoutLogin` which uses `dataBundle` and in this case it will use the `dataBundle` variable updated by overide method but we need to have the `TypicalDataBundle`
So that why its not working. 
I may not be right :sweat_smile:

Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java]: Perhaps we can add a reference `typicalDataBundle` to `BaseActionTest` and use that instead? `dataBundle` will then be used only for data specific to the test.

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java]: @whipermr5 Have a look on current changes :)

File: src/test/java/teammates/test/cases/action/BaseActionTest.java [hunk 2]
```diff
@@ -50,6 +52,8 @@ public void baseClassSetup() {
 
     protected void prepareTestData() {
         dataBundle = getTypicalDataBundle();
+        typicalBundle = getTypicalDataBundle();
+        removeAndRestoreDataBundle(typicalBundle);
         removeAndRestoreTypicalDataBundle();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/BaseActionTest.java]: Now the typical data bundle will be removed and restored twice.

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/BaseActionTest.java]: Done

File: src/test/java/teammates/test/cases/action/BaseActionTest.java [hunk 3]
```diff
@@ -50,7 +52,8 @@ public void baseClassSetup() {
 
     protected void prepareTestData() {
         dataBundle = getTypicalDataBundle();
-        removeAndRestoreTypicalDataBundle();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/BaseActionTest.java]: On second thoughts, it'd be better to move the reference to `BaseComponentTestCase`, as `BaseComponentTestCase` already has `getTypicalDataBundle()` and `removeAndRestoreDataBundle(dataBundle)` in `removeAndRestoreTypicalDataBundle`.

**Comment by sujeet14108** [comment]
@whipermr5 have a look

File: src/test/java/teammates/test/cases/BaseComponentTestCase.java [hunk 4]
```diff
@@ -31,6 +31,7 @@
 
     protected static final GaeSimulation gaeSimulation = GaeSimulation.inst();
     protected static final BackDoorLogic backDoorLogic = new BackDoorLogic();
+    protected DataBundle typicalBundle = getTypicalDataBundle();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/BaseComponentTestCase.java]: With this, the method body of `removeAndRestoreTypicalDataBundle` below can be simplified to one line: `removeAndRestoreDataBundle(typicalBundle)`.

File: src/test/java/teammates/test/cases/action/AdminEmailComposePageActionTest.java [hunk 5]
```diff
@@ -91,7 +93,10 @@ public void testExecuteAndPostProcess() throws Exception {
     }
 
     @Override
+    @Test
     protected void testAccessControl() throws Exception {
-        //TODO: implement this
+        dataBundle = getTypicalDataBundle();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminEmailComposePageActionTest.java]: Do update all the places where you can now just use `typicalBundle`, instead of getting it again.

File: src/test/java/teammates/test/cases/BaseComponentTestCase.java (hunk 4 above)
Comment (whipermr5) [inline src/test/java/teammates/test/cases/BaseComponentTestCase.java]: Also, the method body of `prepareTestData` in `BaseActionTest` can do without the first line.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionCopyActionTest.java [hunk 6]
```diff
@@ -29,10 +29,12 @@ protected void prepareTestData() {
 
     @BeforeMethod
     public void refreshTestData() {
+        super.prepareTestData();
         dataBundle = getTypicalDataBundle();
         removeAndRestoreTypicalDataBundle();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionCopyActionTest.java]: These two lines are unnecessary.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackCopyActionTest.java [hunk 7]
```diff
@@ -28,10 +28,12 @@ protected void prepareTestData() {
 
     @BeforeMethod
     public void refreshTestData() {
+        super.prepareTestData();
         dataBundle = getTypicalDataBundle();
         removeAndRestoreTypicalDataBundle();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackCopyActionTest.java]: These two lines are unnecessary.

File: src/test/java/teammates/test/cases/BaseComponentTestCase.java [hunk 8]
```diff
@@ -97,7 +97,8 @@ protected StudentAttributes getStudent(StudentAttributes student) {
     }
 
     protected void removeAndRestoreTypicalDataBundle() {
-        removeAndRestoreDataBundle(typicalBundle);
+        DataBundle dataBundle = getTypicalDataBundle();
+        removeAndRestoreDataBundle(dataBundle);
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/BaseComponentTestCase.java]: Why didn't `removeAndRestoreDataBundle(typicalBundle);` work?

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/BaseComponentTestCase.java]: I don't know exactly but we not need to reload data as some test are depend on previous test data so  ` removeAndRestoreDataBundle(typicalBundle);`  is not working.
Also `removeAndRestoreTypicalDataBundle()` is used in different other test packages as well so would it be good to change this method ?

Comment (whipermr5) [inline src/test/java/teammates/test/cases/BaseComponentTestCase.java]: Find out why, and fix the actual cause. `removeAndRestoreDataBundle(typicalBundle);` should have no difference from `DataBundle dataBundle = getTypicalDataBundle(); removeAndRestoreDataBundle(dataBundle);`.

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/BaseComponentTestCase.java]: `DataBundle dataBundle = getTypicalDataBundle(); removeAndRestoreDataBundle(tdataBundle);`

In this we are updating the DB with new DataBundle. And its not same as 
`removeAndRestoreDataBundle(typicalBundle);` as `typicalBundle` which is modified in between
test.

**Comment by sujeet14108** [comment]
@whipermr5  have a look on current changes. Can't we do it like this?

**Comment by whipermr5** [comment]
> @whipermr5 have a look on current changes. Can't we do it like this?

I'll take a look :)

File: src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java [hunk 9]
```diff
@@ -335,9 +335,14 @@ protected StudentProfilePictureAction getAction(String... params) {
...
 
+    @Test
     @Override
     protected void testAccessControl() throws Exception {
-        //TODO: implement this
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.USER_ID, account.googleId,
+                Const.ParamsNames.BLOB_KEY, account.studentProfile.pictureKey
+        };
+        verifyAnyRegisteredUserCanAccess(submissionParams);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: Access control for this action seems to be well tested by existing methods. Move relevant methods under `testExecuteAndPostProcess` to `testAccessControl`.

**Comment by whipermr5** [comment]
@sujeet14108 Any updates here?

File: src/test/java/teammates/test/cases/action/InstructorCourseEditSaveActionTest.java [hunk 10]
```diff
@@ -175,6 +175,17 @@ protected String getPageResultDestination(String parentUri, boolean isError, Str
     @Override
     @Test
     protected void testAccessControl() throws Exception {
-        //TODO: implement this
+        InstructorAttributes instructor = typicalBundle.instructors.get("instructor1OfCourse1");
+        String courseId = instructor.courseId;
+        String courseName = CoursesLogic.inst().getCourse(courseId).getName();
```
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/InstructorCourseEditSaveActionTest.java]: Provide directly as 
`String courseName = "Typical Course 1 with 2 Evals";`

File: src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java [hunk 11]
```diff
@@ -22,13 +22,15 @@ protected String getActionUri() {
...
     protected void prepareTestData() {
+        super.prepareTestData();
         dataBundle = loadDataBundle("/InstructorEditInstructorFeedbackPageTest.json");
         removeAndRestoreDataBundle(dataBundle);
     }
 
     @Override
     @Test
     public void testExecuteAndPostProcess() {
+        prepareTestData();
```
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/InstructorEditInstructorFeedbackPageActionTest.java]: Not needed.

**Comment by sujeet14108** [comment]
@whipermr5 `InstructorFeedbackQuestionCopyActionTest` passing on my local

**Comment by whipermr5** [comment]
@sujeet14108 Well it's failing on mine. Try merging the latest master and running it again.

**Comment by whipermr5** [comment]
@sujeet14108 Let us know when you've addressed all of @leeyimin's concerns and this is ready for review :)

File: src/test/java/teammates/test/cases/action/StudentProfileCreateFormUrlActionTest.java [hunk 12]
```diff
@@ -74,9 +74,12 @@ protected StudentProfileCreateFormUrlAction getAction(String... params) {
         return (StudentProfileCreateFormUrlAction) gaeSimulation.getActionObject(getActionUri(), params);
     }
 
+    @Test
     @Override
     protected void testAccessControl() throws Exception {
-        //TODO: implement this
+        String[] submissionParams = new String[] {};
+        verifyAccessibleForStudents(submissionParams);
+        verifyAccessibleForAdminToMasqueradeAsInstructor(submissionParams);
```
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/StudentProfileCreateFormUrlActionTest.java]: I really don't know what to check here :sweat_smile:

**Comment by sujeet14108** [comment]
Please review :)

File: src/test/java/teammates/test/cases/action/InstructorEditStudentFeedbackPageActionTest.java [hunk 13]
```diff
@@ -195,6 +196,20 @@ protected InstructorEditStudentFeedbackPageAction getAction(String... params) {
...
+        String moderatedStudentEmail = student.email;
+
+        String[] submissionParams = new String[]{
+                Const.ParamsNames.COURSE_ID, courseId,
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionName,
+                Const.ParamsNames.FEEDBACK_SESSION_MODERATED_PERSON, moderatedStudentEmail
+        };
+
+        verifyOnlyInstructorsOfTheSameCourseCanAccess(submissionParams);
+        verifyUnaccessibleWithoutModifySessionPrivilege(submissionParams);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorEditStudentFeedbackPageActionTest.java]: `verifyUnaccessibleWithoutModifySessionCommentInSectionsPrivilege` is more appropriate here.

File: src/test/java/teammates/test/cases/action/InstructorEditStudentFeedbackSaveActionTest.java [hunk 14]
```diff
@@ -25,13 +26,16 @@ protected String getActionUri() {
...
+        super.prepareTestData();
         dataBundle = loadDataBundle("/InstructorEditStudentFeedbackPageTest.json");
         removeAndRestoreDataBundle(dataBundle);
     }
 
     @Override
     @Test
     public void testExecuteAndPostProcess() {
+        prepareTestData();
+
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorEditStudentFeedbackSaveActionTest.java]: Is this necessary?

File: src/test/java/teammates/test/cases/action/InstructorEditStudentFeedbackSaveActionTest.java [hunk 15]
```diff
@@ -659,6 +663,19 @@ protected String getPageResultDestination(
...
+        String moderatedStudentEmail = student.email;
+
+        String[] submissionParams = new String[]{
+                Const.ParamsNames.COURSE_ID, courseId,
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionName,
+                Const.ParamsNames.FEEDBACK_SESSION_MODERATED_PERSON, moderatedStudentEmail
+        };
+
+        verifyOnlyInstructorsOfTheSameCourseCanAccess(submissionParams);
+        verifyUnaccessibleWithoutModifyCoursePrivilege(submissionParams);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorEditStudentFeedbackSaveActionTest.java]: `verifyUnaccessibleWithoutModifySessionCommentInSectionsPrivilege` is more appropriate here.

Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorEditStudentFeedbackSaveActionTest.java]: `testDifferentPrivileges` should be moved here.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java [hunk 16]
```diff
@@ -24,13 +24,15 @@ protected String getActionUri() {
...
     protected void prepareTestData() {
+        super.prepareTestData();
         dataBundle = loadDataBundle("/InstructorFeedbackEditCopyTest.json");
         removeAndRestoreDataBundle(dataBundle);
     }
 
     @Override
     @Test
     public void testExecuteAndPostProcess() {
+        prepareTestData();
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: Is this still needed?

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java [hunk 17]
```diff
@@ -413,6 +415,14 @@ protected InstructorFeedbackEditCopyAction getAction(String... params) {
...
-        //TODO: implement this
+        String[] params = new String[]{
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, "First feedback session",
+                Const.ParamsNames.COURSE_ID, "idOfTypicalCourse1",
+                Const.ParamsNames.COPIED_FEEDBACK_SESSION_NAME, "Session with valid name",
+                Const.ParamsNames.COPIED_COURSES_ID, "idOfTypicalCourse2",
+                Const.ParamsNames.COPIED_COURSES_ID, "idOfSampleCourse-demo"
+        };
+
+        verifyUnaccessibleWithoutModifyCoursePrivilege(params);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: InstructorFeedbackEditCopyAction checks for `INSTRUCTOR_PERMISSION_MODIFY_SESSION`, so should use `verifyUnaccessibleWithoutModifySessionPrivilege`.

Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: Should add `verifyOnlyInstructorsOfTheSameCourseCanAccess`.

File: src/test/java/teammates/test/cases/action/StudentProfileCreateFormUrlActionTest.java (hunk 12 above)
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentProfileCreateFormUrlActionTest.java]: I think just `verifyAccessibleForStudents` is sufficient here.

File: src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java [hunk 18]
```diff
@@ -335,9 +334,10 @@ protected StudentProfilePictureAction getAction(String... params) {
         return (StudentProfilePictureAction) gaeSimulation.getActionObject(getActionUri(), params);
     }
 
+    @Test
     @Override
     protected void testAccessControl() throws Exception {
-        //TODO: implement this
+        testActionWithEmailAndCourse();
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: In `testActionWithEmailAndCourse`, only `testActionWithEmailAndCourseUnauthorisedInstructorOrStudent` and `testActionWithEmailAndCourseUnauthorisedInstructorOrStudentMasquerade` are more relevant to testing of access control. So `testActionWithEmailAndCourse` should not be moved entirely to `testAccessControl`.

**Comment by sujeet14108** [comment]
@leeyimin Please review :)

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java [hunk 19]
```diff
@@ -413,6 +414,14 @@ protected InstructorFeedbackEditCopyAction getAction(String... params) {
...
-        //TODO: implement this
+        String[] params = new String[]{
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, "First feedback session",
+                Const.ParamsNames.COURSE_ID, "idOfTypicalCourse1",
+                Const.ParamsNames.COPIED_FEEDBACK_SESSION_NAME, "Session with valid name",
+                Const.ParamsNames.COPIED_COURSES_ID, "idOfTypicalCourse2",
+                Const.ParamsNames.COPIED_COURSES_ID, "idOfSampleCourse-demo"
+        };
+
+        verifyUnaccessibleWithoutModifySessionPrivilege(params);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: You should try to move lines 101 - 137 here.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackRemindParticularStudentsActionTest.java [hunk 20]
```diff
@@ -75,8 +75,15 @@ protected InstructorFeedbackRemindParticularStudentsAction getAction(String... p
...
     }
 
+    @Test
     @Override
     protected void testAccessControl() throws Exception {
-        //TODO: implement this
+        FeedbackSessionAttributes fs = typicalBundle.feedbackSessions.get("session1InCourse1");
+        String[] submissionParams = new String[]{
+                Const.ParamsNames.COURSE_ID, fs.getCourseId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, fs.getFeedbackSessionName()
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackRemindParticularStudentsActionTest.java]: Add param for userlist so that the params are valid.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackRemindParticularStudentsActionTest.java [hunk 21]
```diff
@@ -75,8 +75,15 @@ protected InstructorFeedbackRemindParticularStudentsAction getAction(String... p
...
+        };
+
+        verifyOnlyInstructorsOfTheSameCourseCanAccess(submissionParams);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackRemindParticularStudentsActionTest.java]: Can add `verifyUnaccessibleWithoutModifySessionPrivilege`.

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java [hunk 22]
```diff
@@ -78,10 +78,18 @@ protected InstructorStudentRecordsAjaxPageAction getAction(String... params) {
...
-      //TODO: implement this
+        InstructorAttributes instructor = typicalBundle.instructors.get("instructor3OfCourse1");
+        StudentAttributes student = typicalBundle.students.get("student2InCourse1");
+
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, instructor.courseId,
+                Const.ParamsNames.STUDENT_EMAIL, student.email,
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, "First feedback session"
+        };
+        verifyOnlyInstructorsOfTheSameCourseCanAccess(submissionParams);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: Add `verifyUnaccessibleWithoutViewStudentInSectionsPrivilege`.

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java [hunk 23]
```diff
@@ -78,10 +78,18 @@ protected InstructorStudentRecordsAjaxPageAction getAction(String... params) {
...
-    @Override
     @Test
+    @Override
     protected void testAccessControl() throws Exception {
-      //TODO: implement this
+        InstructorAttributes instructor = typicalBundle.instructors.get("instructor3OfCourse1");
+        StudentAttributes student = typicalBundle.students.get("student2InCourse1");
+
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, instructor.courseId,
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: Why are you accessing `courseId` from `instructor`?

Comment (sujeet14108) [reply src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: I have seen in above function like line 57 where it is done like this only. Also student has no attribute of courseid

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java (hunk 22 above)
Comment (sujeet14108) [reply src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: This check is failing with `java.lang.AssertionError: expected:<You are not registered in the course > but was:<>`

Comment (leeyimin) [reply src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: Can you add a test case to ensure that resultTables of InstructorStudentRecordsAjaxPageData is empty for the helper instructor?

Comment (leeyimin) [reply src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: Actually, the test case I wanted you to add is above in testExecuteAndPostProcess, maybe you could move that test case here instead.

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java (hunk 23 above)
Comment (leeyimin) [reply src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: It is probably better to retrieve the course from the typicalBundle and use the course id from the course, rather than a random instructor of the course.

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java (hunk 22 above)
Comment (sujeet14108) [reply src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: Done

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java (hunk 23 above)
Comment (sujeet14108) [reply src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: Have a look

**Comment by sujeet14108** [comment]
@leeyimin Ready for review

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java [hunk 24]
```diff
@@ -48,17 +49,40 @@ public void testExecuteAndPostProcess() {
...
+        CourseAttributes course = typicalBundle.courses.get("typicalCourse1");
+
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, course.getId(),
+                Const.ParamsNames.STUDENT_EMAIL, student.email,
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, "First feedback session"
+        };
+        verifyOnlyInstructorsOfTheSameCourseCanAccess(submissionParams);
+
         ______TS("Typical case: instructor cannot view sections");
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: May want to give appropriate description here.

**Comment by leeyimin** [comment]
@sujeet14108 Try to move lines 101 - 137 to `testAccessControl` for InstructorFeedbackEditCopyActionTest.java .

**Comment by sujeet14108** [comment]
@leeyimin In InstructorFeedbackEditCopyActionTest  lines 101 - 137 is using
` dataBundle = loadDataBundle("/InstructorFeedbackEditCopyTest.json");`
So to move it in access control we have to change the bundle

**Comment by leeyimin** [comment]
You can still access `dataBundle` from `testAccessControl`. Take note to do `gaeSimulation.loginAsInstructor()` as the correct instructor.

**Comment by sujeet14108** [comment]
@leeyimin Have a look

**Comment by sujeet14108** [comment]
Will suppress that unused variable

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java [hunk 25]
```diff
@@ -413,6 +378,62 @@ protected InstructorFeedbackEditCopyAction getAction(String... params) {
...
+
+        ______TS("Failure case: copying from course with insufficient permission");
+        params = new String[] {
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, fs.getFeedbackSessionName(),
+                Const.ParamsNames.COURSE_ID, "FeedbackEditCopy.CS2107",
+                Const.ParamsNames.COPIED_FEEDBACK_SESSION_NAME, "valid name",
+                Const.ParamsNames.COPIED_COURSES_ID, course.getId()
+        };
+        InstructorFeedbackEditCopyAction a = getAction(params);
+        @SuppressWarnings("unused")
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: Do not suppress the warning, it is not necessary. You can just execute the action instead using `a.executeAndPostProcess()`.

**Comment by sujeet14108** [comment]
@leeyimin have a look

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java (hunk 19 above)
Comment (leeyimin) [reply src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: Done.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java [hunk 26]
```diff
@@ -413,6 +376,58 @@ protected InstructorFeedbackEditCopyAction getAction(String... params) {
...
+
+        ______TS("Failure case: copying from course with insufficient permission");
+        params = new String[] {
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, fs.getFeedbackSessionName(),
+                Const.ParamsNames.COURSE_ID, "FeedbackEditCopy.CS2107",
+                Const.ParamsNames.COPIED_FEEDBACK_SESSION_NAME, "valid name",
+                Const.ParamsNames.COPIED_COURSES_ID, course.getId()
+        };
+        InstructorFeedbackEditCopyAction a = getAction(params);
+        String expectedString;
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: Declare `expectedString` at line 409 and line 428 instead so that the code looks cleaner.

File: src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java [hunk 27]
```diff
@@ -48,17 +49,40 @@ public void testExecuteAndPostProcess() {
...
+        CourseAttributes course = typicalBundle.courses.get("typicalCourse1");
+
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, course.getId(),
+                Const.ParamsNames.STUDENT_EMAIL, student.email,
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, "First feedback session"
+        };
+        verifyOnlyInstructorsOfTheSameCourseCanAccess(submissionParams);
+
+        ______TS("Typical case: instructor cannot view sections without View-Student-In-Sections privilege");
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/InstructorStudentRecordsAjaxPageActionTest.java]: Maybe remove the phrase `Typical case`.

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java (hunk 26 above)
Comment (sujeet14108) [reply src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: Will do 
Forgot to save file before pushing :p

**Comment by sujeet14108** [comment]
@wkurniawan07 have a look
